code/game/objects/items/devices/radio/listening_bugs.dm

```diff
@@ -0,0 +1,298 @@
+	/// Whether or not the bug can be used to listen to its own channel.
+	var/prevent_snooping = FALSE
+
+/obj/item/device/radio/listening_bug/ui_data(mob/user)
+	var/list/data = list()
+
+	data["broadcasting"] = broadcasting
+	data["listening"] = listening
+	data["frequency"] = frequency
+	data["freqlock"] = freqlock
+
+	var/list/radio_channels = list()
+
+	for(var/channel in channels)
+		var/channel_key = channel_to_prefix(channel)
+		radio_channels += list(list(
+			"name" = channel,
+			"status" = channels[channel] & FREQ_LISTENING,
+			"hotkey" = channel_key))
+
+	data["channels"] = radio_channels
+
+	data["command"] = volume
+	data["useCommand"] = use_volume
+	data["subspace"] = subspace_transmission
+	data["subspaceSwitchable"] = subspace_switchable
+	data["headset"] = FALSE
+
+	return data
+
+/obj/item/device/radio/listening_bug/ui_act(action, params, datum/tgui/ui, datum/ui_state/state)
+	switch(action)
+		if("listen")
+			if(prevent_snooping)
+				to_chat(usr, SPAN_WARNING("This device cannot receive transmissions!"))
+				return
+			listening = !listening
+			return
+		if("subspace")
+			if(!ishuman(usr))
+				return
+			var/mob/living/carbon/human/user = usr
+			if(!check_access(user.wear_id) && !check_access(user.get_active_hand()))
+				to_chat(user, SPAN_WARNING("You need an authenticated ID card to change this function!"))
+				return
+			if(subspace_switchable)
+				subspace_transmission = !subspace_transmission
+				if(!subspace_transmission)
+					channels = list()
+				return
+	..()
+
+/obj/item/device/radio/listening_bug/hear_talk(mob/M as mob, msg, verb = "says", datum/language/speaking = null)
+	var/processed_verb = "[SPAN_RED("\[LSTN Device\]")] [verb]"
+	if(broadcasting)
+		if(get_dist(src, M) <= 7)
+			talk_into(M, msg,null,processed_verb,speaking)
+
+/obj/item/device/radio/listening_bug/verb/change_disguise()
+	set name = "Change Disguise"
+	set category = "Object"
+	set src in usr
+
+	if(usr.is_mob_incapacitated())
+		to_chat(usr, SPAN_WARNING("You cannot do this while incapacitated!"))
+		return FALSE
+
+	var/check = tgui_alert(usr, "Do you wish to change the disguise of this listening bug?", "Change Disguise?", list("Yes", "No"))
+	if(check != "Yes")
+		return FALSE
+
+	var/list/camo_options = list(
+		DISGUISE_REMOVE,DISGUISE_RADIO,DISGUISE_PEN,DISGUISE_FOUNTAIN_PEN,DISGUISE_ACCESS_TUNER,DISGUISE_WHISTLE,DISGUISE_MASS_SPEC,DISGUISE_CAMERA,DISGUISE_ZIPPO,DISGUISE_TAPE_RECORDER
+	)
+	camo_options -= current_disguise
+	var/new_disguise = tgui_input_list(usr, "What new disguise do you wish to use?", "New Disguise", camo_options)
+
+	if(!new_disguise || (new_disguise == current_disguise))
+		return FALSE
+
+	handle_new_disguise(new_disguise)
+	return TRUE
+
+/obj/item/device/radio/listening_bug/proc/handle_new_disguise(new_disguise)
+	if(!new_disguise)
+		return FALSE
+	var/new_icon
+	var/new_name
+	var/new_desc
+	var/new_icon_state
+	var/new_item_state
+	var/new_equip_slots
+	switch(new_disguise)
```

Comment:
I'm not sure this is the best way to do what I want here.

Reply:
use initial